source/MaterialXCore/Node.cpp

```diff
@@ -252,14 +252,28 @@ vector<ElementPtr> NodeGraph::topologicalSort() const
 
     const vector<ElementPtr>& children = getChildren();
 
-    // Calculate in-degrees for all nodes, and enqueue those with degree 0.
-    std::unordered_map<ElementPtr, size_t> inDegree(children.size());
+    // Calculate in-degrees for all children.
+    std::unordered_map<ElementPtr, int> inDegree(children.size());
     std::deque<ElementPtr> childQueue;
     for (ElementPtr child : children)
     {
-        NodePtr node = child->asA<Node>();
-        inDegree[child] = node ? node->getDownstreamPorts().size() : 0;
-        if (inDegree[child] == 0)
+        int connectionCount = 0;
+        if (child->isA<Output>())
+        {
+            connectionCount += int(!child->asA<Output>()->getNodeName().empty());
+        }
+        else
+        {
+            for (InputPtr input : child->getChildrenOfType<Input>())
+            {
+                connectionCount += int(!input->getNodeName().empty());
+            }
+        }
+
+        inDegree[child] = connectionCount;
```

Comment:
Since an Edge can be implicitly cast to a boolean, here's one way this logic could be written:

```
int connectionCount = 0;
for (int i = 0; i < child->getUpstreamEdgeCount(); i++)
{
    if (child->getUpstreamEdge(i))
    {
        connectionCount++;
    }
}
inDegree[child] = connectionCount;
```